internal/discovery/namespace_filter.go

```diff
@@ -0,0 +1,94 @@
+package discovery
+
+import (
+	"github.com/newrelic/nri-kubernetes/v3/internal/config"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	listersv1 "k8s.io/client-go/listers/core/v1"
+)
+
+// NamespaceFilterer provides an interface to filter namespaces.
+type NamespaceFilterer interface {
+	IsAllowed(namespace string) bool
+}
+
+// NamespaceFilter is a struct holding pointers to the config and the namespace lister.
+type NamespaceFilter struct {
+	c      *config.Config
+	lister listersv1.NamespaceLister
+}
+
+// NewNamespaceFilter inits the namespace lister and returns a new NamespaceFilter.
+func NewNamespaceFilter(c *config.Config, client kubernetes.Interface, options ...informers.SharedInformerOption) (*NamespaceFilter, chan<- struct{}) {
+	stopCh := make(chan struct{})
+
+	factory := informers.NewSharedInformerFactoryWithOptions(client, defaultResyncDuration, options...)
+
+	lister := factory.Core().V1().Namespaces().Lister()
+
+	factory.Start(stopCh)
+	factory.WaitForCacheSync(stopCh)
+
+	return &NamespaceFilter{
+		c:      c,
+		lister: lister,
+	}, stopCh
+}
+
+// IsAllowed checks given any namespace, if it's allowed to be scraped by using the NamespaceLister
+func (nf *NamespaceFilter) IsAllowed(namespace string) bool {
```

Comment:
As discussed, I wonder if it would be possible to not only used the cached list of namespaces but also the `IsAllowed` results for each namespace. However, it probably doesn't worth it to look into it further because it may be quite complex and just a premature optimization.

Reply:
Yes, the thing is that adding this cache maintained by us would add extra complexity and taking into account the amount of namespaces a big cluster can have (some hundreds?) the clients would barely notice the difference in performance; maybe we'd be over engineering a bit.

Anyway, a good point to bring up. Let's see if anyone has something to add about this.

Reply:
Added the cache layer here 3f558b7cb9c5a01a8b5fff641506b2131c8fab68